crates/testing/src/view_sync_task.rs

```diff
@@ -0,0 +1,202 @@
+use async_compatibility_layer::channel::UnboundedStream;
+use futures::FutureExt;
+use hotshot_task::task::{HotShotTaskCompleted, HotShotTaskTypes};
+use hotshot_task::{
+    event_stream::ChannelStream,
+    task::{FilterEvent, HandleEvent, HandleMessage, TS},
+    task_impls::{HSTWithEventAndMessage, TaskBuilder},
+    MergeN,
+};
+use hotshot_task_impls::events::HotShotEvent;
+use hotshot_types::traits::node_implementation::{NodeType, TestableNodeImplementation};
+use snafu::Snafu;
+use std::{
+    collections::{HashMap, HashSet},
+    sync::Arc,
+};
+
+use crate::{test_launcher::TaskGenerator, test_runner::Node, GlobalTestEvent};
+
+/// ViewSync Task error
+#[derive(Snafu, Debug, Clone)]
+pub struct ViewSyncTaskErr {
+    hit_view_sync: HashSet<usize>,
+}
+
+/// ViewSync task state
+pub struct ViewSyncTask<TYPES: NodeType, I: TestableNodeImplementation<TYPES>> {
+    /// the node handles
+    pub(crate) handles: Vec<Node<TYPES, I>>,
+    /// nodes that hit view sync
+    pub(crate) hit_view_sync: HashSet<usize>,
+}
+
+impl<TYPES: NodeType, I: TestableNodeImplementation<TYPES>> TS for ViewSyncTask<TYPES, I> {}
+
+/// ViewSync task types
+pub type ViewSyncTaskTypes<TYPES, I> = HSTWithEventAndMessage<
+    ViewSyncTaskErr,
+    GlobalTestEvent,
+    ChannelStream<GlobalTestEvent>,
+    (usize, HotShotEvent<TYPES>),
+    MergeN<UnboundedStream<HotShotEvent<TYPES>>>,
+    ViewSyncTask<TYPES, I>,
+>;
+
+#[derive(Clone, Debug, Copy)]
+pub enum ShouldHitViewSync {
+    /// the node should hit view sync
+    Yes,
+    /// the node should not hit view sync
+    No,
+    /// don't care if the node should hit view sync
+    Ignore,
+}
+
+/// Description for a view sync task.
+#[derive(Clone, Debug)]
+pub enum ViewSyncTaskDescription {
+    /// (min, max) number nodes that may hit view sync, inclusive
+    Threshold(usize, usize),
+    /// node idx -> whether or not the node should hit view sync
+    /// if node not in map, assumed to be `ShouldHItViewSync::DontCare`
+    Precise(HashMap<usize, ShouldHitViewSync>),
```

Comment:
We can probably simplify this a little bit.  I think either all nodes should hit view sync or none.  The only exception here is the fact that some nodes will be down.  I think this can be just simplified to `Threshold` and then eliminate some of the matching below.